Set custom color when applying initial background image

draganescu opened the pull request:
Closes #44167

The original attempt was using `colorThief` (#22564) and there is another PR (#46908) which I learned after doing this one  doing the same thing.

Anyway 😬 this PR aims to land the feature.



https://github.com/WordPress/gutenberg/assets/107534/62be4171-4549-46eb-859f-e61e88ce02e4

ajlende commented on packages/block-library/src/cover/shared.js:
```diff
@@ -200,3 +203,30 @@ export async function getCoverIsDark( url, dimRatio = 50, overlayColor ) {
+			return color.hex;
+		} catch ( error ) {
+			// If there's an error, just assume the image is dark.
+			return false;
```
If there's a CORS error, this will return `#FFF` (from the `defaultColor`); whereas, other types of errors will return `false` which I'm guessing should be `undefined` (so it doesn't set the attribute to `false`) where the CSS default is `#000`.

draganescu replied:
In https://github.com/WordPress/gutenberg/pull/54054/commits/dc9d046ffdbe56a93525ffee2b5d6937ab66e8e9 I've set the default to be #000 and return undefined if there is an error. Not sure how to approach otherwise?

ajlende reviewed (changes requested):
The `isDark` calculation uses the cover color to figure out if the text should be light or dark. So setting this should also set `isDark`.

draganescu commented:
@ajlende added the dark check _after_ the media sets custom color.

ajlende commented on packages/block-library/src/cover/shared.js:
```diff
@@ -73,12 +73,24 @@ export function attributesFromMedia( setAttributes, dimRatio ) {
+
+		// Only pass the url to getCoverIsDark if the media is an image as video is not handled.
+		const newUrl = media?.type === 'image' ? media.url : undefined;
+		const isDark = await getCoverIsDark(
+			newUrl,
+			dimRatio,
+			customOverlayColor
+		);
```
We should use the same URL for both of them. If I remember correctly, we're only running FastAverageColor on images for performance.

ajlende replied:
Also, we could speed things up by only running FastAverageColor once and using the result for both the `overlayColor` _and_ `isDark` attributes.

draganescu replied:
yeah we could but the refactoring brrr :)

ajlende replied:
We're both reducing code duplication with this refactoring and improving performance. I think it's worth it.

draganescu commented:
@ajlende did some refactoring to use retrieving average color functions once. Got rid of an `useEffect` 🎉 and seems more lean in general. I hope nothing breaks 😁

ajlende commented on packages/block-library/src/cover/shared.js:
```diff
@@ -156,47 +160,66 @@ function retrieveFastAverageColor() {
 				// Previously the default color was white, but that changed
 				// in v6.0.0 so it has to be manually set now.
-				defaultColor: [ 255, 255, 255, 255 ],
+				defaultColor: [ 0, 0, 0, 255 ],
 				// Errors that come up don't reject the promise, so error
 				// logging has to be silenced with this option.
 				silent: process.env.NODE_ENV === 'production',
 				crossOrigin: imgCrossOrigin,
 			} );
 			// FAC uses 0-255 for alpha, but colord expects 0-1.
-			const media = { r, g, b, a: a / 255 };
-			const composite = compositeSourceOver( overlay, media );
-			return colord( composite ).isDark();
+			return color;
```
In many cases FAC logs errors and returns the `defaultColor` instead of throwing errors<sup>[1](https://github.com/fast-average-color/fast-average-color/blob/51e8b27fec3dd5e5474162c6e88fc56135079dce/src/index.ts#L97-L100),[2](https://github.com/fast-average-color/fast-average-color/blob/51e8b27fec3dd5e5474162c6e88fc56135079dce/src/index.ts#L107-L110),[3](https://github.com/fast-average-color/fast-average-color/blob/51e8b27fec3dd5e5474162c6e88fc56135079dce/src/index.ts#L117-L119),[4](https://github.com/fast-average-color/fast-average-color/blob/51e8b27fec3dd5e5474162c6e88fc56135079dce/src/index.ts#L126-L129),[5](https://github.com/fast-average-color/fast-average-color/blob/51e8b27fec3dd5e5474162c6e88fc56135079dce/src/index.ts#L152-L157)</sup>. So those errors return black, but other errors that they don't catch return `null`. I wonder if we should just return the same `defaultColor` on the errors that we catch since we already can't differentiate between some errors and a black image.

Also, since the `defaultColor` switched to black, the behavior here deviates from `getCoverIsDark` which [assumes a white default](https://github.com/WordPress/gutenberg/pull/54054/files#diff-98e8fbed8e10d45fee8dfcb0373d733aa1335f7cbf0a5dd883739b86fae0fd7eR189). I wonder if those should be the same?

draganescu replied:
In `getCoverIsDark` the default is white because we want to see how much it darkens it. I don't think they're related. I can't say we should return black for all errors, it makes little sense. Plus we should then return the full color object FAC builds, but we';'d have to simulate it hence we'll have to always keep it up to date. I'd stick to null.

ajlende replied:
> Plus we should then return the full color object FAC builds, but we';'d have to simulate it hence we'll have to always keep it up to date. 

Or it could return a `Colord` or `RgbaColor` object since we're already using `colord` everywhere else.

ajlende replied:
> In `getCoverIsDark` the default is white because we want to see how much it darkens it. I don't think they're related. 

In one case we're assuming the `overlayColor` goes on white and the other black.

The white default existed because FAC was hardcoded to white at one point. And I think `getCoverIsDark` was mistakenly changed to return `true` by default in #53961—it should have been `false` to match the FAC default.

ajlende commented on packages/block-library/src/cover/edit/index.js:
```diff
@@ -97,6 +98,8 @@ function CoverEdit( {
 		tagName: TagName = 'div',
 	} = attributes;
 
+	const [ averageMediaColor, setAverageMediaColor ] = useState( undefined );
```
Because you're storing the `averageMediaColor` with `useState`, when you reload the page, the state is lost. Changing the `dimRatio`, for example, after that results in incorrect calculations since `averageMediacolor` is then `undefined`.

draganescu replied:
Oh no another effect to write.

ajlende replied:
Or a function to memoize :)

ajlende commented on packages/block-library/src/cover/shared.js:
```diff
@@ -73,12 +73,16 @@ export function attributesFromMedia( setAttributes, dimRatio ) {
 			mediaType = media.type;
 		}
 
+		const isDark = getCoverIsDark( dimRatio, averageBackgroundColor?.hex );
```
It looks like `attributesFromMedia` is only used when changing the media. And if you are _always_ setting the `customOverlayColor` to the `averageBackgroundColor`, we can skip the `compositeSourceOver` operation. It will _always_ be the same as the `averageBackgroundColor` because you're overlaying the same color on top of itself now.

draganescu replied:
Am I not doing that by providing the 3rd argument? One `compositeSourceOver` will need to happen on a white background to see how dark the average is.

ajlende replied:
> Am I not doing that by providing the 3rd argument? One `compositeSourceOver` will need to happen on a white background to see how dark the average is.

It's not quite the same. Overlaying a 50% opaque red over white results in pink (current behavior), but a 50% opaque red over red results in red (correct behavior).

https://github.com/WordPress/gutenberg/pull/54054/files#diff-98e8fbed8e10d45fee8dfcb0373d733aa1335f7cbf0a5dd883739b86fae0fd7eR187-R191

ajlende commented on packages/block-library/src/cover/shared.js:
```diff
@@ -156,47 +160,66 @@ function retrieveFastAverageColor() {
  * calculating the relative darkness of the Cover.
  *
- * @param {string} url
  * @param {number} dimRatio
  * @param {string} overlayColor
+ * @param {Object} averageBackgroundColor
  * @return {Promise<boolean>} True if cover should be considered to be dark.
```
I don't see anything `async` in here, so this no longer returns a `Promise`.

```suggestion
 * @param {number} dimRatio
 * @param {string} overlayColor
 * @param {Object} averageBackgroundColor
 * @return {boolean} True if cover should be considered to be dark.
```

ajlende commented on packages/block-library/src/cover/shared.js:
```diff
@@ -156,47 +160,66 @@ function retrieveFastAverageColor() {
+export function getCoverIsDark(
+	dimRatio = 50,
+	overlayColor,
+	averageBackgroundColor = undefined
```
`= undefined` is redundant here.

```suggestion
	averageBackgroundColor
```

ajlende commented on packages/block-library/src/cover/shared.js:
```diff
@@ -156,47 +160,66 @@ function retrieveFastAverageColor() {
+ * This method evaluates the average color of a background image
+ *
+ * @param {string} url
+ * @return {object|null} Null the color can't be computed, or the color.
```
This actually returns a `Promise`.

```suggestion
 * @return {Promise<Object|null>} Null the color can't be computed, or the color.
```

draganescu commented:
I am starting over this one.

draganescu commented:
There are a few things to try and achieve with this:

1. The main objective: to set the overlay color as the average media color for image backgrounds
2. The refactoring objectives:
2.1 To make sure for the same image background we compute the average media color once
2.2 To reuse the average background media color in all the computation of the is dark attribute

(2.1) and (2.2) make refactoring annoying because:
- the cover block accrued a lot of features and there are many possible situations
- the average color computation is an async operation which is annoying to memoize
- using a state is not immediately manageable in an effect because of the two ways in which we use the color (hex to store, rgb to compute composite)

draganescu committed:
set overlay from image background average color and refactor code to compute it only once per URL and set is dark attribute correctly

ajlende commented on packages/block-library/src/cover/edit/index.js:
```diff
@@ -59,6 +61,82 @@ function getInnerBlocksTemplate( attributes ) {
+ * 									 combination composite result is dark
+ */
+function computeIsDark( dimRatio, overlayColor, backgroundColor ) {
+	/**
+	 * Performs a Porter Duff composite source over operation on two rgba colors.
+	 *
+	 * @see https://www.w3.org/TR/compositing-1/#porterduffcompositingoperators_srcover
+	 *
+	 * @param {import('colord').RgbaColor} source Source color.
+	 * @param {import('colord').RgbaColor} dest   Destination color.
+	 * @return {import('colord').RgbaColor} Composite color.
+	 */
+	function compositeSourceOver( source, dest ) {
+		return {
+			r: source.r * source.a + dest.r * dest.a * ( 1 - source.a ),
+			g: source.g * source.a + dest.g * dest.a * ( 1 - source.a ),
+			b: source.b * source.a + dest.b * dest.a * ( 1 - source.a ),
+			a: source.a + dest.a * ( 1 - source.a ),
+		};
+	}
```
This is a pure function, so I don't see any reason that we have to re-create it every time we call `computeIsDark`. This should be a top-level function like before.

ajlende commented on packages/block-library/src/cover/edit/index.js:
```diff
@@ -59,6 +61,82 @@ function getInnerBlocksTemplate( attributes ) {
+ * @return {Promise<string>} average color in hex
+ */
+async function computeAverageColor( url ) {
+	function retrieveFastAverageColor() {
+		if ( ! retrieveFastAverageColor.fastAverageColor ) {
+			retrieveFastAverageColor.fastAverageColor = new FastAverageColor();
+		}
+		return retrieveFastAverageColor.fastAverageColor;
+	}
```
We aren't getting the benefits of caching the FAC instance if we're rebuilding the function every time we call `comuteAverageColor`. This should be a top-level function like before.

ajlende commented on packages/block-library/src/cover/edit/index.js:
```diff
@@ -59,6 +61,82 @@ function getInnerBlocksTemplate( attributes ) {
+		return color.hex;
+	} catch ( error ) {
+		// If there's an error return the fallback color.
+		return '#FFFFFF';
```
```suggestion
		return DEFAULT_AVERAGE_COLOR;
```

draganescu replied:
This should be `DEFAULT_AVERAGE_COLOR`

ajlende commented on packages/block-library/src/cover/edit/index.js:
```diff
@@ -146,27 +231,44 @@ function CoverEdit( {
+		return computedAverageColor;
+	};
+
+	// Set the overlay color to the average color of th
+	// featured image and and is dark based on that color.
+	const updateFromFeaturedImage = async () => {
+		if ( ! useFeaturedImage ) {
+			return;
+		}
+		await setOverlayFromAverageColor( {
+			url: mediaUrl,
+			type: IMAGE_BACKGROUND_TYPE,
+		} );
+		setIsDark(
+			overlayColor.color,
+			await getAverageBackgroundColor( mediaUrl )
+		);
+	};
+	// we call this on render since the url is
+	// 'hijacked' by the featured image
+	updateFromFeaturedImage();
```
This is causing an effect, so it needs to be wrapped in `useEffect`.

ajlende commented on packages/block-library/src/cover/edit/index.js:
```diff
@@ -114,25 +192,32 @@ function CoverEdit( {
+		__unstableMarkNextChangeAsNotPersistent();
+		const color = await getAverageBackgroundColor( newUrl );
+		setOverlayColor( color );
+		setIsDark( color, color );
```
You don't have to run the composite source over operation on this since the colors are the same. `setAttributes( { isDark: colord( color ).isDark() } )` would work just as well.

ajlende commented on packages/block-library/src/cover/edit/index.js:
```diff
@@ -146,27 +231,44 @@ function CoverEdit( {
+		await setOverlayFromAverageColor( newMedia );
 	};
 
 	const onClearMedia = async () => {
```
I don't think this has to be `async` anymore.

ajlende commented on packages/block-library/src/cover/edit/index.js:
```diff
@@ -296,8 +389,8 @@ function CoverEdit( {
 				: undefined,
-			isDark: isDarkSetting,
 		} );
+		setIsDark( overlayColor.color, await getAverageBackgroundColor( url ) );
```
Moving the await below the other `setAttributes` is going to cause the attributes to be update in two steps with a delay vs one step before. I haven't run the code yet, but I expect this will trigger two renders instead of one.

The effect of everything else being updated and then the `isDark` setting lagging behind will be more obvious on slower hardware. Moving the `await` to the top will probably have the same behavior as before, but I'm not 100% certain because `setAttributes` is weird and it would still get called twice, just closer together.

Same thing for `onSetOverlayColor` and `onUpdateDimRatio`.

dmsnell replied:
also good to think about making atomic updates so that state is never partially updated and thus inconsistent. we definitely only want one undo/redo point for this operation, if any at all (which I guess we might not want an undo step since this flows from the other user-initiated action?)

ajlende replied:
I updated some places, but I had to draw the line somewhere with all the refactoring that was required. And the line I drew was refactoring out `withColors` which is a HOC that adds the `setOverlayColor` prop which calls `setAttributes`.

I made sure that everywhere that `setOverlayColor` was followed by another `setAttributes`, there was an `__unstableMarkNextChangeAsNotPersistent()` call to keep the undo levels to what you would expect while editing.

draganescu committed:
don't reset the maually set overlay, stop redefining utility functions, improve featured image case, don't compute darkness for same color background and foreground

draganescu commented:
Addressed @ajlende latest review and fixed the behavior to respect the manually set color overlay on load. Also tweaked the use featured image behavior.

Updated the OP with a video with current working state.

Following a conversation with @richtabor there is the question of when should the overlay be set from the average color of the background image? All the time when a background image is applied or only if there is no overlay color? Or only when the overlay color is not manually set?

Ideally we would know the difference between an overlay color manually set by a user and one set by the system. And in this PR we do, but only until the page is reloaded.

The best simplification would be to set the overlay from average color of background image only if there is no overlay color and no background image.

Otherwise we'd have to save a new attribute, so that when the user switches the background image we can tell if the color is the one from the previous image average color, or if it's set manually, and on media change only suggest if the color is set by the system.

We need to decide between:
1. leaving it like it is in the PR (on all media changes system will set a new overlay)
2. system setting an overlay only when the block is not setup with overlay or background
3. adding a `manualOverlay` flag to set so that system sets overlay only when overlay is set previously by syetem

richtabor commented:
> adding a manualOverlay flag to set so that system sets overlay only when overlay is set previously by system. 

I think this is the correct path, though it can be in a follow-up to get this in. 

I would expect an overlay color suggested as long as I didn’t manually change the color myself. If I set an overlay color, that color should persist on media change.

draganescu committed:
update test with the new default color being white

ajlende commented:
> adding a `manualOverlay` flag to set so that system sets overlay only when overlay is set previously by syetem.

I wonder if we could use the synchronous FAC API on save when an `overlayColor` isn't set? Not sure if it would work or if that would cause problems, but it would keep the block API cleaner rather than adding a new attribute.

ajlende commented:
> I wonder if we could use the synchronous FAC API on save when an `overlayColor` isn't set? Not sure if it would work or if that would cause problems, but it would keep the block API cleaner rather than adding a new attribute.

This got me wondering why `isDark` doesn't compute on save either. Turns out that `withColors` causes problems when trying to do this. Only the slug for a preset color is saved in the attributes, so it's a whole mess to get the actual color from the slug in `save`.

For the average color, we can't run the synchronous API because of how we're serializing the content from the `save` function. It requires an `HTMLImageElement` which we don't have access to in `save`.

---

Do you think we should add a block deprecation for blocks that previously had the default black overlay color so they keep the black overlay when the image for the block is changed? Or will it be fine that replacing the image on these old blocks starts computing the color from now on?

draganescu commented:
> Do you think we should add a block deprecation for blocks that previously had the default black overlay color so they keep the black overlay when the image for the block is changed? Or will it be fine that replacing the image on these old blocks starts computing the color from now on?

I think if the image is replace and the user didn't set any color that the system acting as designed is expected behavior. The user relied on a default and the default is updated. If they change the image and don't like the default they can always use black from the palette.

draganescu committed:
add the userOverlayColor attribute to keep track of system vs user color setting

ajlende committed:
Remove default value to fix integration tests

ajlende committed:
Update JSDoc comments

ajlende committed:
Consolidate all color functions into color-utils.js

dmsnell reviewed (commented):
I'm not sure if I'm missing something, but I don't see the intended effect demonstrated in the preview video on the PR. Maybe the network connection is too fast so that the image loads before the background color can show through? Would be nice to see this running where it has a visual impact.

Will leave another note here for whatever it's worth; I see that much of this code is shuffling around existing code, but there are two ways in which this seems like a good candidate for two additional asynchronous opportunities:
 - can we lazy-load these color utilities, or at least separate the functions like `computeAverageColor` (which pull in the image-processing dependencies) into a separate file where it will be more trivial later to lazy-load them?
 - can we move some of this computation into a WebWorker or other worker where it won't block the main thread? in many cases I can imagine this introducing a notable delay; consider someone uploading a 30 MB JPG image from their new phone with a high-end camera. even on fast algorithms that can take some time. if we can move this to a worker then I think all we'd need to send is a URL and all we'd need to receive is a short color string.

ajlende committed:
Memoize the average color calculation with memize

ajlende committed:
Move await before setAttributes calls so they won't be spread out

ajlende committed:
Consolidate setAttributes calls

ajlende committed:
Refactor out setIsDark

ajlende committed:
Refactor attributesFromMedia to return the attributes rather than set them

ajlende commented:
> can we lazy-load these color utilities, or at least separate the functions like `computeAverageColor` (which pull in the image-processing dependencies) into a separate file where it will be more trivial later to lazy-load them?

I didn't think our build process supported code splitting with dynamic imports to do this; does it? If it's supported, that would be wonderful! There's lots of places I think it would be useful.

> can we move some of this computation into a WebWorker or other worker where it won't block the main thread? in many cases I can imagine this introducing a notable delay; consider someone uploading a 30 MB JPG image from their new phone with a high-end camera. even on fast algorithms that can take some time. if we can move this to a worker then I think all we'd need to send is a URL and all we'd need to receive is a short color string.

I agree it's a worthwhile optimization, but not for this PR. I've been testing with an 83MB 8k (7680x4320) image which is even slow enough that it takes half a second on my M1 MacBook Pro. The fast-average-color API is pretty bad, and I figured it could be replaced at some point when there's a better library. At that time, it would be good to get this running in a web worker.

ajlende committed:
Refactor out setOverlayFromAverageColor to merge more setAttribute calls

ajlende committed:
Refactor compositeIsDark to return early if second params are the same

ajlende committed:
Rename next* to new* for consistency with other existing functions

ajlende committed:
Move useEffect up near related code

ajlende committed:
Fix comment

ajlende committed:
Add check for fully opaque overlay colors

ajlende committed:
Fix clearMedia from not resetting overlayColor when not manually specified

ajlende committed:
Rename getAverageBackgroundColor to getMediaColor

ajlende committed:
Fix use newDimRatio on media select

ajlende committed:
Fix overlay color not resetting when clearing featured image

ajlende committed:
Fix missing averageBackgroundColor when toggling featured image

ajlende committed:
Rename userOverlayColor to isUserOverlayColor

ajlende committed:
Fix isDark calculation when clearing media

ajlende committed:
Fix undo states

ajlende committed:
Update docs with new attribute

ajlende committed:
Separate out compositeIsDark calls from setAttributes

ajlende committed:
Fix case where isDark wasn't correctly set when toggling off featured image

ajlende committed:
Move useFeaturedImage: false to index

ajlende committed:
Move focalPoint: undefined to index

ajlende committed:
Set useFeaturedImage to undefined instead of false for slightly cleaner attributes

ajlende committed:
Shorten #FFFFFF to #FFF

ajlende committed:
Fix overlayColor in onUpdateDimRatio

ajlende commented on packages/block-library/src/cover/edit/index.js:
```diff
@@ -141,72 +156,113 @@ function CoverEdit( {
-		const isDarkSetting = await getCoverIsDark(
-			undefined,
+	const onClearMedia = () => {
+		let newOverlayColor = overlayColor.color;
+		if ( ! isUserOverlayColor ) {
+			newOverlayColor = DEFAULT_OVERLAY_COLOR;
+			setOverlayColor( undefined );
+
+			// Make undo revert the next setAttributes and the previous setOverlayColor.
+			__unstableMarkNextChangeAsNotPersistent();
+		}
+
```
When I was playing around with test cases, it felt more natural for the overlay color to be reset when clearing the media if the user hadn't specified their own overlay color.

1. Create a cover block with featured image or media. (Do not manually set an overlay color.)
2. Clear the media.
3. What should happen?

If it should revert to using the default CSS black overlay, nothing needs to be changed.

However, if you think it should retain the computed color even with `isUserOverlayColor` unset, then you can apply this suggestion.

```suggestion
```

draganescu replied:
I think this is a good catch in terms of UX - if the system set the color based on an image that is _removed_ so should the color be removed.

ajlende committed:
Fix use newDimRatio for isDark on updateDimRatio

ajlende commented on packages/block-library/src/cover/edit/index.js:
```diff
@@ -285,18 +341,42 @@ function CoverEdit( {
+			? averageBackgroundColor
+			: overlayColor.color;
+
+		if ( ! isUserOverlayColor ) {
+			if ( newUseFeaturedImage ) {
+				setOverlayColor( newOverlayColor );
+			} else {
+				setOverlayColor( undefined );
+			}
+
+			// Make undo revert the next setAttributes and the previous setOverlayColor.
+			__unstableMarkNextChangeAsNotPersistent();
+		}
```
Similar to clearing media that I mentioned [above](https://github.com/WordPress/gutenberg/pull/54054#discussion_r1330883005), when I was playing around with test cases, it felt more natural for the overlay color to be reset when toggling off the featured image if the user hadn't specified their own overlay color.

1. Create cover block with featured image. (Do not manually set an overlay color.)
2. Toggle off the featured image.
3. What should happen?

If it should revert to using the default CSS black overlay, nothing needs to be changed.

However, if you think it should retain the computed color even with `isUserOverlayColor` unset, then you can apply this suggestion.

```suggestion
		if ( ! isUserOverlayColor && newUseFeaturedImage) {
			setOverlayColor( newOverlayColor );

			// Make undo revert the next setAttributes and the previous setOverlayColor.
			__unstableMarkNextChangeAsNotPersistent();
		}
```

draganescu replied:
Same as the case of setting media, using a featured image with a system set color, means removing featured image should remove the color.

ajlende commented on packages/block-library/src/cover/block.json:
```diff
@@ -42,6 +42,9 @@
 		"customOverlayColor": {
 			"type": "string"
 		},
+		"isUserOverlayColor": {
+			"type": "boolean"
+		},
```
I switched this to have an 'is' prefix so it wouldn't be confused with an actual color.

draganescu replied:
Yes, good naming tweak thank you!

ajlende committed:
Fix use newDimRatio for isDark on toggleUseFeaturedImage

ajlende committed:
Rename DEFAULT_AVERAGE_COLOR to DEFAULT_BACKGROUND_COLOR to better describe what it means in the contexts used

ajlende committed:
Better JSDoc comment

ajlende committed:
Fix DEFAULT_BACKGROUND_COLOR import

ajlende commented on packages/block-library/src/cover/edit/index.js:
```diff
@@ -141,72 +156,113 @@ function CoverEdit( {
 			isRepeated: undefined,
-			useFeaturedImage: false,
-			isDark: isDarkSetting,
+			useFeaturedImage: undefined,
+			isDark: newIsDark,
```
I've removed the `isUserOverlayColor: false` from here because the behavior was unexpected to me.

1. Set a custom color on an existing cover block with media.
2. Clear the media.
3. Add new media.
4. What should happen?

If the color I set in step 1 should be used, nothing needs to be changed. This is what I expected.

However, if you think that the new media should have its average color computed even though we set a custom color in step 1, then you can apply this suggestion.

```suggestion
			isUserOverlayColor: undefined,
			useFeaturedImage: undefined,
			isDark: newIsDark,
```

draganescu replied:
Yes this too is in line with all the notes above: if there is a user color we should keep it and also keep the flag since the user set the overlay manually.

ajlende reviewed (approved):
I've changed the behavior in a few places to better match my expectations when using the feature. Each place is marked with a review comment. If you wish to restore any of that behavior, I've added suggestions to those comments that can be applied.

And I've tested this pretty extensively and fixed most of the things I've found. However, I've found one thing that I didn't have time to fix.

Existing blocks using a featured image _and_ a custom overlay color get their custom overlay color overridden by the `useEffect` when the post is loaded. The only way I can think to solve this is to add a block deprecation that sets `isUserOverlayColor` when `( overlayColor || customOverlayColor ) && useFeaturedImage )` is true.

This seems minor enough to handle in a follow-up so we can get this into 6.4, so it has my approval. 👍

ajlende commented:
Looks like all the failing tests here are also failing on trunk. Someone with more permissions than I have will have to merge this.

t-hamano commented:
I also believe that the test failure was not caused by this PR, but it may have been caused by a problem occurring with GitHub Actions. They [tweeted](https://x.com/githubstatus/status/1704428675406446976?s=20) that this issue has been resolved, so I'll try running the test again.

ajlende committed:
Update tests to use playwright toHaveCSS

ajlende committed:
Variable naming

ajlende committed:
Replace remaining instance of checking backgrounds with toHaveCSS

ajlende committed:
Update image upload test

ajlende committed:
Update image transform test

ajlende committed:
Add block deprecation for auto overlay color

ajlende committed:
Fix lint

ajlende committed:
Remove unnecessary expect for the default value

ajlende commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -39,18 +38,14 @@ test.describe( 'Cover', () => {
+			'background-color',
+			'rgb(0, 0, 0)'
+		);
 	} );
```
When I was refactoring this test to use `toHaveCSS` like the other ones, I saw that it wasn't really testing what it said in the description, so I just fixed it in the process.

ajlende commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -200,7 +194,7 @@ test.describe( 'Cover', () => {
 	} );
 
-	test( 'dims the background image down by 50% when transformed from the Image block', async ( {
+	test( 'dims the background image down by 50% black when transformed from the Image block', async ( {
```
I specified black here because it isn't possible to run FAC in a transformation due to it being asynchronous and it may have been confusing that it isn't getting the average color like the other test.

ajlende reviewed (approved):
> Existing blocks using a featured image _and_ a custom overlay color get their custom overlay color overridden by the `useEffect` when the post is loaded. The only way I can think to solve this is to add a block deprecation that sets `isUserOverlayColor` when `( overlayColor || customOverlayColor ) && useFeaturedImage )` is true.

I fixed this issue by adding the block deprecation and decided to do it for anything that had a color set since that's how things would be set now.

---

Also, I think I fixed the tests to not be flakey anymore. I found the [`_expect` function](https://github.com/microsoft/playwright/blob/main/packages/playwright-core/src/server/frames.ts#L1366) that `toHaveCSS` uses, and it looks like it does retry until the expect returns success or a max timeout is reached, so it should be reliable now even though we're asynchronously setting the overlay color.

---

@draganescu I'll leave you to take another look at everything in case I missed something since I'll be out for the next few days. Hopefully the tests pass 🤞

draganescu committed:
update mobile snapshots due to new attribute

draganescu committed:
updated snapshots for transforms on mobile tests due to new attribute
